src/Application.vala

```diff
@@ -103,12 +103,57 @@ public class Music.Application : Gtk.Application {
         settings.bind ("window-maximized", main_window, "maximized", SettingsBindFlags.SET);
     }
 
+    // FIXME(aitor-gomila): improve algorithm and make code cleaner
+    Array<File> listDirectory(string directory) {
+        var dir = Dir.open(directory, 0);
+        string? name = null;
+        Array<File> elements = new Array<File>();
+
+        while((name = dir.read_name ()) != null) {
+            string filePath = Path.build_filename (directory, name);
+            File file = File.new_for_path(filePath);
+            elements.append_val(file);
+        }
+        return elements;
+    }
+
+    private bool isDirectory(string file) {
+        return FileUtils.test(file, FileTest.IS_DIR);
+    }
+
+    private void loopThroughFiles(Array<File> files) {
+        for (var i = 0; i < files.length; i++) {
+            var file = files.index(i);
+            var filePath = file.get_path();
+            var isDirectory = isDirectory(filePath);
+
+            if (!isDirectory) {
+                File[] arr = new File[1];
+                arr[0] = file;
+                playback_manager.queue_files(arr);
+                return;
```

Comment:
You shouldn't return here. User may have opened many files, not just one.

Reply:
The loop should go through all files, return should only stop the current iteration for the current file. What should I use instead?

Reply:
Use `continue`